cpp/src/arrow/array/builder_base.h

```diff
@@ -170,9 +170,14 @@ class ARROW_EXPORT ArrayBuilder {
     if (new_capacity < 0) {
       return Status::Invalid("Resize capacity must be positive");
     }
+// gcc is aggressively complaining about overflows (due to Resize and
+// GrowByFactor). Overflowing int64_t is an indication of other problems.
```

Comment:
Perhaps not the focus of this PR, but gcc is probably right in this case. Signed overflow is undefined behaviour, so one cannot reliably detect overflows with `new_capacity < old_capacity`. You either have to check beforehand (by doing a subtraction rather than an add) or do the whole resizing computation in unsigned land.

Reply:
I can't find the source invocation causing this.

Reply:
I suspect it's code like:
```
    auto new_length = length_ + length;
    ARROW_RETURN_NOT_OK(CheckCapacity(new_length, length_));
```

Though it's not easy to second-guess the compiler, especially without seeing the actual error message :-)

Reply:
```
In file included from ../src/arrow/array/builder_adaptive.h:22:0,
                 from ../src/arrow/builder.h:22,
                 from ../src/arrow/csv/converter.cc:26:
../src/arrow/array/builder_base.h: In member function ‘virtual arrow::Status arrow::csv::{anonymous}::NullConverter::Convert(const arrow::csv::BlockParser&, int32_t, std::shared_ptr<arrow::Array>*)’:
../src/arrow/array/builder_base.h:174:5: error: assuming signed overflow does not occur when assuming that (X + c) < X is always false [-Werror=strict-overflow]
     if (new_capacity < old_capacity) {
     ^~
```

Reply:
Tried commenting a few places similar to what you suggest.

Reply:
Ok found it.

Reply:
I suppose it's:
```cpp
    ARROW_RETURN_NOT_OK(CheckCapacity(length_ + 1, length_));
```
in `NullBuilder::AppendNull`.

I suspect that solving this without silencing the warning will involve a fair amount of casting...